### PR TITLE
[ruby] add `METHOD_REF` and `TYPE_REF` nodes for methods and class/module decls

### PR DESCRIPTION
**Pull request by AndreiDreyer**
This PR handles:
 * Added `METHOD_REF` nodes directly under `:program` for all methods defined in file
 * Added `TYPE_REF` nodes directly under `:program` for all class/modules defined in file

### CODE REVIEW

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 1]
```diff
@@ -94,14 +96,71 @@ class AstCreator(
...
+    val typeNameForMethods = scope.surroundingTypeFullName
+      .map { x =>
+        x.split("[:]{2}").dropRight(1).mkString("")
+      }
+      .getOrElse(Defines.Undefined)
+
+    programSummary.namespaceToType
+      .filter(_._1 == typeNameForMethods)
+      .flatMap(_._2)
+      .filter(x => x.name.contains(":program"))
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:120]: Could this be translated into non-`StubbedType`s?

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 2]
```diff
@@ -94,14 +96,71 @@ class AstCreator(
...
+      .flatMap(_.methods)
+      .filterNot(x => x.name.contains("<init>") || x.name.contains("<clinit>"))
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala]: Use the `x2cpg.Defines` contains here

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 3]
```diff
@@ -94,14 +96,71 @@ class AstCreator(
...
         scope.popScope()
         methodAst(methodNode_, Seq.empty, bodyAst, methodReturn, newModifierNode(ModifierTypes.MODULE) :: Nil)
       }
       .getOrElse(Ast())
   }
+
+  private def methodRefNodesForInternalMethods(): List[Ast] = {
+    val typeNameForMethods = scope.surroundingTypeFullName
+      .map { x =>
+        x.split("[:]{2}").dropRight(1).mkString("")
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:113]: Is this not just `stripSuffix(Defines.Program)`?

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 4]
```diff
@@ -44,6 +44,8 @@ class AstCreator(
       .map(_.stripPrefix(java.io.File.separator))
       .getOrElse(fileName)
 
+  private val internalLineAndColNum: Option[Integer] = Option(1)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:47]: I'd feel more comfortable if these were fresh values for each node
```suggestion
  private def internalLineAndColNum: Option[Integer] = Option(1)
```

Comment (DavidBakerEffendi) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:47]: But if this is what Pythonsrc does then we can leave it at `val`

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 5]
```diff
@@ -94,14 +96,71 @@ class AstCreator(
...
+          .columnNumber(internalLineAndColNum)
+
+        astForAssignment(methodRefIdent, methodRefNode, internalLineAndColNum, internalLineAndColNum)
+      }
+      .toList
+  }
+
+  private def typeRefNodesForInternalDecls(): List[Ast] = {
+    programSummary.namespaceToType
+      .filter(_._1.contains(scope.surroundingTypeFullName.getOrElse(Defines.Undefined)))
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:144]: We don't want this to execute if `scope.surroundingTypeFullName` is `None`. So I recommend, at the top, starting with `scope.surroundingTypeFullName.foreach`, e.g.
```scala 
scope.surroundingTypeFullName.map { surroundingTypeFullName => 
   programSummary.namespaceToType /* etc. */
}.getOrElse(List.empty)
```

**Review by DavidBakerEffendi — APPROVED** [review]

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala (hunk 2 above)
Comment (AndreiDreyer) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala]: Removed this `filterNot`, these methods are already being filtered out during the above `filter(_._1. ==)`

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala (hunk 1 above)
Comment (DavidBakerEffendi) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:120]: I meant filtering based on `StubbedType` trait

**Review by DavidBakerEffendi — APPROVED** [review]

**Pull request merged**